pkg/kubelet/kubeletconfig/util/files/files_test.go

```diff
@@ -0,0 +1,383 @@
+			expectedErr:  fmt.Errorf(""), // should have error, but the error info can only be fulfilled after creating the file
+		},
+		"exists as dir": {
+			path:         prePath + "/ExistsAsDir",
+			isFile:       false,
+			create:       true,
+			expectedBool: true,
+			expectedErr:  nil,
+		},
+	}
+
+	for testcaseName, testcase := range testcases {
+		var err error
+		t.Run(testcaseName, func(t *testing.T) {
+			if testcase.create {
+				if testcase.isFile {
+					err = createFile(fs, testcase.path)
+				} else {
+					err = createDir(fs, testcase.path)
+				}
+			}
+
+			// not nil indicates that we should specify the error
+			if testcase.expectedErr != nil {
+				testcase.expectedErr = fmt.Errorf("expected dir at %q, but mode is %q", testcase.path, getFileMode(fs, testcase.path))
+			}
+
+			if err != nil {
+				t.Errorf("unexpected error in creating %q, err: %v", testcase.path, err)
+			} else {
+				if result, err := DirExists(fs, testcase.path); err != nil {
+					// if the path is file, should get error
+					if !reflect.DeepEqual(testcase.expectedErr, err) {
+						t.Errorf("unexpected error in creating dir, expected: %v, actual: %v", testcase.expectedErr, err)
+					}
+				} else {
+					if testcase.expectedBool != result {
+						t.Errorf("unexpected result, expected: %v, actual: %v", testcase.expectedBool, result)
+					}
+				}
+			}
+		})
+
+	}
+}
+
+func TestEnsureDir(t *testing.T) {
+	prePath := getTempDir(t, "TestEnsureDir")
+	fs := utilfs.DefaultFs{}
+	defer cleanup(t, fs, prePath)
+
+	testcases := map[string]struct {
+		path        string // path to be created
+		isFile      bool   // if the path indicates a file
+		create      bool   // if should create dir before test
+		expectedErr error
+	}{
+		"exists as file": {
+			path:        prePath + "/ExistsAsFile",
+			isFile:      true,
+			create:      true,
+			expectedErr: fmt.Errorf(""), // should have error, but the error info can only be fulfilled after creating the file
+		},
+		"exists as dir": {
+			path:        prePath + "/NotExists",
+			isFile:      false,
+			create:      true,
+			expectedErr: nil,
+		},
+		"not exists in sub dir": {
+			path:        prePath + "/SubDir1/SubDir2/Exists",
+			isFile:      false,
+			create:      false,
+			expectedErr: nil,
+		},
+	}
+
+	for testcaseName, testcase := range testcases {
+		var err error
+		t.Run(testcaseName, func(t *testing.T) {
+			if testcase.create {
+				if testcase.isFile {
+					err = createFile(fs, testcase.path)
+				} else {
+					err = createDir(fs, testcase.path)
+				}
+			}
+
+			// not nil indicates we should specify the error
+			if testcase.expectedErr != nil {
+				testcase.expectedErr = fmt.Errorf("expected dir at %q, but mode is %q", testcase.path, getFileMode(fs, testcase.path))
+			}
+
+			if err != nil {
+				t.Errorf("unexpected error in creating dir: %s, error: %v", testcase.path, err)
+			} else {
+				// ensure dir exists
+				if err = EnsureDir(fs, testcase.path); err != nil {
```

Comment:
use `!reflect.DeepEqual` instead of `err != nil` like you did above?

Reply:
no, otherwise you would find that the "exists as file" test case will go into the else part, which is not right for the logic.